hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/placement/CSMappingPlacementRule.java

```diff
@@ -181,6 +181,10 @@ private void setupGroupsForVariableContext(VariableContext vctx, String user)
       return;
     }
     Set<String> groupsSet = groups.getGroupsSet(user);
+    if (groupsSet.isEmpty()) {
+      LOG.warn("There are no groups for user {}", user);
+      return;
```

Comment:
Just to be sure the variable context is as consistent as possible, please add a 
`vctx.putExtraDataset("groups", groupsSet);`
before the return, empty set is a valid groups, at least it can be used by the matchers, not that it will ever match, but it's perhaps better than leaving it blank.
As for primary and secondary groups there is not much we can do.

Reply:
Added it. Thank you @shuzirra